dotnet/src/Microsoft.Agents.AI.Hosting.OpenAI/Responses/Common/Id/DefaultIdGenerator.cs

```diff
@@ -0,0 +1,153 @@
+    /// <param name="partitionKeyLength">The length of the partition key if generating a new one.</param>
+    /// <param name="infix">Optional additional text to insert between the prefix and the entropy.</param>
+    /// <param name="watermark">Optional text to insert in the middle of the entropy string for traceability.</param>
+    /// <param name="delimiter">The delimiter character used to separate parts of the ID.</param>
+    /// <param name="partitionKey">An explicit partition key to use. When provided, this value will be used instead of generating a new one.</param>
+    /// <param name="partitionKeyHint">An existing ID to extract the partition key from. When provided, the same partition key will be used instead of generating a new one.</param>
+    /// <returns>A new ID with format "{prefix}{delimiter}{infix}{entropy}{delimiter}{partitionKey}".</returns>
+    /// <exception cref="ArgumentException">Thrown when the watermark contains non-alphanumeric characters.</exception>
+    private static string NewId(string prefix, int stringLength = 32, int partitionKeyLength = 16, string infix = "",
+        string watermark = "", string delimiter = "_", string? partitionKey = null, string partitionKeyHint = "")
+    {
+        var entropy = SecureEntropy(stringLength);
+        var pKey = partitionKey ?? (string.IsNullOrEmpty(partitionKeyHint)
+            ? SecureEntropy(partitionKeyLength)
+            : ExtractPartitionId(partitionKeyHint));
+
+        if (!string.IsNullOrEmpty(watermark))
+        {
+            if (!WatermarkRegex().IsMatch(watermark))
+            {
+                throw new ArgumentException($"Only alphanumeric characters may be in watermark: {watermark}",
+                    nameof(watermark));
+            }
+
+            entropy = $"{entropy[..(stringLength / 2)]}{watermark}{entropy[(stringLength / 2)..]}";
+        }
+
+        infix ??= "";
+        prefix = !string.IsNullOrEmpty(prefix) ? $"{prefix}{delimiter}" : "";
+        return $"{prefix}{infix}{entropy}{pKey}";
+    }
+
+    /// <summary>
+    /// Generates a secure random alphanumeric string of the specified length.
+    /// </summary>
+    /// <param name="stringLength">The desired length of the random string.</param>
+    /// <returns>A random alphanumeric string.</returns>
+    /// <exception cref="ArgumentException">Thrown when stringLength is less than 1.</exception>
+    private static string SecureEntropy(int stringLength)
+    {
+        ArgumentOutOfRangeException.ThrowIfLessThan(stringLength, 1);
+        var entropy = "";
+        while (entropy.Length != stringLength)
+        {
+            var buffer = RandomNumberGenerator.GetBytes(stringLength);
+            var entropyBase64 = Convert.ToBase64String(buffer);
+            entropy = new string(entropyBase64.Where(char.IsLetterOrDigit).ToArray())[..stringLength];
+        }
+
+        return entropy;
+    }
```

Comment:
Is this only used on .NET Core? If so, it can just use RandomNumberGenerator.GetString, e.g.
```csharp
private static string SecureEntropy(int stringLength) =>
    RandomNumberGenerator.GetString("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789", stringLength);
```